### PR TITLE
Fix missing color in legend for size + color legends.

### PR DESCRIPTION
**Pull request by haldenl**
There is currently a bug wherein redundantly encoding size + color creates a legend w/ no color.

#4664

image

[editor](https://vega.github.io/editor/#/url/vega-lite/N4KABGBEAkDODGALApgWwIaQFxUQFzwAdYsB6UgN2QHN0A6agSz0QFcAjOxge1IRQyUa6ALQAbZskoBmOgCtY3AHaQANOCgATZAgBOjQnh4qckAIJgE6Asl2Ex3PJcTcA7oyXUwL3bGSE3WzB0JU0wVEYxHTBCINoxByVYMAAzbl0wCnR9blZk+GzYOjUNSE1rTBxQCAhIVl0xbC0K0gLfeUUVDQBfdVqMXQBrJsh4Rl14KJLa5CV4bk0PaibqmsgADxWNGqgUxmQxTRGAWUidAH1Y3XOAcXQE5WmdqDwAT1iRgEdWEKM8a0YVCeO0g7A8TTwulYyG2YF6sMgsEYAC9kFtnrt9ocRgAJdJ+AKuWzAtZvD6mb6-ZgAoF9Z6QdDUai6YR4NGmdBUXSMtGw+Eg+YOXTo+l7A5HUx43z+QLCukgsnsqCUpR-Glo+VrRnM1lKhlcnmQPmayCvEUgsXY0wAeX0TBUmtqipGSm4ESU9yNO26PRA3SAA)

Culprit is the `fill` attribute in `update` created by the `size` legend:

```
{
  ...
  "legends": [
    {
      "fill": "color",
      "gradientLength": {"signal": "clamp(height, 64, 200)"},
      "title": "Average of Horsepower",
      "size": "size",
      "encode": {
        "symbols": {
          "update": {
            "fill": {"value": "black"},
            "shape": {"value": "circle"},
            "fillOpacity": {"value": 1},
            "stroke": {"value": "transparent"}
          }
        }
      }
    }
  ]
}
```

Proposed fix is to remove that `update.fill` property if top-level `fill` is already defined.

### CODE REVIEW

File: src/compile/legend/assemble.ts [hunk 1]
```diff
@@ -26,13 +26,18 @@ export function assembleLegends(model: Model): VgLegend[] {
...
+
     if (legend.encode && legend.encode.symbols) {
       const out = legend.encode.symbols.update;
       if (out.fill && out.fill['value'] !== 'transparent' && !out.stroke && !legend.stroke) {
         // For non color channel's legend, we need to override symbol stroke config from Vega config if stroke channel is not used.
         out.stroke = {value: 'transparent'};
       }
+
+      if (legend.fill && legend.size) {
+        // If fill and size are both used, for non color channel's legend, we need remove fill, as it is already specified.
```
Comment (domoritz) [inline src/compile/legend/assemble.ts]: Shouldn't we always delete this when fill is defined?

Comment (haldenl) [reply src/compile/legend/assemble.ts]: Wasn't sure if under other conditions you may want an override, but I guess that does not make much sense. Fixed.

File: examples/compiled/interactive_seattle_weather.vg.json [hunk 2]
```diff
@@ -562,14 +562,7 @@
...
       "symbolType": "circle",
-      "encode": {
-        "symbols": {
-          "update": {
-            "fill": {"value": "transparent"},
-            "opacity": {"value": 0.7}
-          }
-        }
-      },
+      "encode": {"symbols": {"update": {"opacity": {"value": 0.7}}}},
```
Comment (haldenl) [inline examples/compiled/interactive_seattle_weather.vg.json]: @domoritz Aha point symbol needs the override.

Comment (haldenl) [reply examples/compiled/interactive_seattle_weather.vg.json]: or is this correct? The bar chart layer triggers the fill. Should the legend have circles (to match bar chart) or points (to match scatterplot)?

[example](https://vega.github.io/editor/#/examples/vega-lite/interactive_seattle_weather)

Comment (domoritz) [reply examples/compiled/interactive_seattle_weather.vg.json]: We have to decide. I don't remember what the default resolution is (first or last legend when merging). @kanitw?

Comment (kanitw) [reply examples/compiled/interactive_seattle_weather.vg.json]: What do you mean override here?

File: src/compile/legend/assemble.ts [hunk 3]
```diff
@@ -26,13 +26,18 @@ export function assembleLegends(model: Model): VgLegend[] {
...
-    // For non color channel's legend, we need to override symbol stroke config from Vega config
+
     if (legend.encode && legend.encode.symbols) {
       const out = legend.encode.symbols.update;
       if (out.fill && out.fill['value'] !== 'transparent' && !out.stroke && !legend.stroke) {
         // For non color channel's legend, we need to override symbol stroke config from Vega config if stroke channel is not used.
         out.stroke = {value: 'transparent'};
       }
+
+      if (legend.fill) {
```
Comment (kanitw) [inline src/compile/legend/assemble.ts]: This solves the problem, but I'm wonder if fill shouldn't even be in the component in the first place. (Should we fix this in the parse stage, not here?)

File: examples/compiled/interactive_seattle_weather.vg.json (hunk 2 above)
Comment (haldenl) [reply examples/compiled/interactive_seattle_weather.vg.json]: The `fill` attribute in `... encode.symbols.update` overriding the `fill` attribute at the top level `legend`.

File: src/compile/legend/assemble.ts (hunk 3 above)
Comment (haldenl) [reply src/compile/legend/assemble.ts]: I think the issue is only with multiple channels sharing the same domain, and merging these occurs in this stage.

**Comment by domoritz** [comment]
This should have high priority, I think.

File: examples/compiled/interactive_seattle_weather.vg.json (hunk 2 above)
Comment (haldenl) [reply examples/compiled/interactive_seattle_weather.vg.json]: btw if we decide to remove this override, we probably want to make sure the opacity of the fill matches the opacity of the stroke.

**Comment by domoritz** [comment]
@haldenl What's left here or is this ready?

**Comment by kanitw** [comment]
I’m not convinced if this is the right fix. Please don’t merge yet.

**Comment by domoritz** [comment]
@kanitw @haldenl We should find a solution for this issue. What's the proposed approach? Can we get a simple fix in soon?

**Comment by kanitw** [comment]
Ok I take a closer look.  There is no way to get rid of that "black" entirely given Vega default symbolBaseFill is transparent.   So this PR is fine.

**Comment by haldenl** [comment]
@domoritz @kanitw 

Follow up question: just wanted to make sure we made a decision on the fill of the legend. 

image

The stroke and fill don't quite match here—any thoughts?

**Comment by domoritz** [comment]
Is that coming from opacity on the fill?

**Comment by haldenl** [comment]
yep

**Comment by domoritz** [comment]
We have a default mark opacity that might be applied here. I have to see the full spec.

**Comment by haldenl** [comment]
It's the weather example

[https://vega.github.io/editor/#/examples/vega-lite/interactive_seattle_weather](https://vega.github.io/editor/#/examples/vega-lite/interactive_seattle_weather)

**Pull request merged**